File: 06-DevOps/05-IaC与配置管理/Terraform生命周期元参数与图变换/selfcheck_tflifecycle.py
# -*- coding: utf-8 -*-
"""tflifecycle 自检：全部基于实际读过的官方原文口径。"""
from tflifecycle import (Resource, propagate_cbd, plan, destroy_graph, has_cycle,
                         _update_or_replace)

N = 0
FAIL = []


def check(label, cond, detail=""):
    global N
    N += 1
    if not cond:
        FAIL.append("%s  %s" % (label, detail))


def run(state, config, deps):
    a, e, cbd, order, trig = plan(state, config, deps)
    return a, e, cbd, order, trig


# ---- 1. 基础动作：无 lifecycle ----
cfg = {"aws_instance.web": Resource("aws_instance.web", {"ami": "ami-2"})}
a, e, cbd, order, _ = run({"aws_instance.web": {"ami": "ami-1"}}, cfg, {})
check("A1 普通属性变化是 update", a == {"aws_instance.web": "update"} and not e, a)
check("A2 默认非 CBD", cbd["aws_instance.web"] is False, cbd)

# 需要替换的属性
cfg = {"aws_instance.web": Resource("aws_instance.web", {"ami": "ami-2"},
                                    requires_replace=["ami"])}
a, e, cbd, order, _ = run({"aws_instance.web": {"ami": "ami-1"}}, cfg, {})
check("A3 不可原地改的属性→replace", a == {"aws_instance.web": "replace"}, a)
check("A4 默认替换次序是先销毁后新建",
      order["aws_instance.web"] == ("destroy", "create"), order)

# ---- 2. create_before_destroy ----
cfg = {"aws_instance.web": Resource("aws_instance.web", {"ami": "ami-2"},
                                    requires_replace=["ami"],
                                    lifecycle={"create_before_destroy": True})}
a, e, cbd, order, _ = run({"aws_instance.web": {"ami": "ami-1"}}, cfg, {})
check("B1 CBD 显式置位", cbd["aws_instance.web"] is True, cbd)
check("B2 CBD 替换次序反过来",
      order["aws_instance.web"] == ("create", "destroy"), order)

# ---- 3. CBD 沿依赖边传播给被依赖方 ----
cfg = {
    "aws_instance.web": Resource("aws_instance.web", {"ami": "ami-2"},
                                 requires_replace=["ami"],
                                 lifecycle={"create_before_destroy": True}),
    "aws_eip.ip": Resource("aws_eip.ip", {}),
}
deps = {"aws_instance.web": ["aws_eip.ip"]}
eff, conflicts = propagate_cbd(cfg, deps)
check("C1 依赖方被隐式置位", eff["aws_eip.ip"] is True, eff)
check("C2 未显式写 false 时无冲突", conflicts == [], conflicts)

# 传递性：A → B → C
cfg3 = {
    "a": Resource("a", {}, lifecycle={"create_before_destroy": True}),
    "b": Resource("b", {}),
    "c": Resource("c", {}),
}
eff, _ = propagate_cbd(cfg3, {"a": ["b"], "b": ["c"]})
check("C3 传播是传递的", eff["b"] and eff["c"], eff)
eff, _ = propagate_cbd(cfg3, {"b": ["a"]})  # 反向：b 依赖 a
check("C4 不会反向传播给依赖方", eff["a"] is True and eff["b"] is False, eff)

# 显式 false 与隐式 true 冲突（官方：会隐含依赖图中的环）
cfg4 = {
    "a": Resource("a", {}, lifecycle={"create_before_destroy": True}),
    "b": Resource("b", {}, lifecycle={"create_before_destroy": False}),
}
eff, conflicts = propagate_cbd(cfg4, {"a": ["b"]})
check("C5 显式 false 撞上隐式 true 记为冲突", conflicts == ["b"], conflicts)
a, e, cbd, order, _ = run({}, cfg4, {"a": ["b"]})
check("C6 冲突会进 errors", any("成环" in x for x in e), e)

# ---- 4. prevent_destroy ----
cfg = {"aws_db_instance.db": Resource("aws_db_instance.db", {"size": "large"},
                                      requires_replace=["size"],
                                      lifecycle={"prevent_destroy": True})}
a, e, cbd, order, _ = run({"aws_db_instance.db": {"size": "small"}}, cfg, {})
check("D1 prevent_destroy 拒绝替换", "aws_db_instance.db" in a and any("prevent_destroy" in x for x in e), e)
a, e, cbd, order, _ = run({"aws_db_instance.db": {"size": "small"}},
                          {"aws_db_instance.db": Resource("aws_db_instance.db",
                                                          {"size": "small"},
                                                          lifecycle={"prevent_destroy": True})}, {})
check("D2 无变化时放行", a == {"aws_db_instance.db": "noop"} and not e, (a, e))
# 官方：配置整体移除时不阻止
a, e, cbd, order, _ = run({"aws_db_instance.db": {"size": "small"}}, {}, {})
check("D3 配置移除后照常 destroy", a == {"aws_db_instance.db": "destroy"}, a)
cfg = {"aws_db_instance.db": Resource("aws_db_instance.db", {"size": "large"},
                                      requires_replace=["size"],
                                      lifecycle={"prevent_destroy": True})}
a, e, cbd, order, _ = run({}, cfg, {})
check("D4 仅创建时不触发 prevent_destroy", a == {"aws_db_instance.db": "create"} and not e, (a, e))

# ---- 5. ignore_changes ----
cfg = {"aws_instance.web": Resource("aws_instance.web", {"tags": "b"},
                                    lifecycle={"ignore_changes": ["tags"]})}
a, e, cbd, order, _ = run({"aws_instance.web": {"tags": "a"}}, cfg, {})
check("E1 被忽略的属性不产生 update", a == {"aws_instance.web": "noop"}, a)
cfg = {"aws_instance.web": Resource("aws_instance.web", {"tags": "b", "ami": "x"},
                                    lifecycle={"ignore_changes": ["tags"]})}
a, e, cbd, order, _ = run({"aws_instance.web": {"tags": "a", "ami": "y"}}, cfg, {})
check("E2 只忽略列出的属性", a == {"aws_instance.web": "update"}, a)
cfg = {"aws_instance.web": Resource("aws_instance.web", {"tags": "b"},
                                    lifecycle={"ignore_changes": "all"})}
a, e, cbd, order, _ = run({"aws_instance.web": {"tags": "a"}}, cfg, {})
check("E3 all 时永不 update", a == {"aws_instance.web": "noop"}, a)
# 官方：create 时 ignore_changes 不起作用
a, e, cbd, order, _ = run({}, cfg, {})
check("E4 ignore_changes 不影响 create", a == {"aws_instance.web": "create"}, a)
# 官方只写明「create 考虑、update 忽略」，未规定「需要替换」这一分支。
# 本模型口径：被忽略的属性不进入 diff，故既不 update 也不 replace（README 已标注口径）。
cfg = {"aws_instance.web": Resource("aws_instance.web", {"ami": "b"},
                                    requires_replace=["ami"],
                                    lifecycle={"ignore_changes": ["ami"]})}
a, e, cbd, order, _ = run({"aws_instance.web": {"ami": "a"}}, cfg, {})
check("E5 模型口径：忽略项不进 diff 故不触发 replace",
      a == {"aws_instance.web": "noop"}, a)

# ---- 6. replace_triggered_by ----
cfg = {
    "aws_ecs_service.svc": Resource("aws_ecs_service.svc", {"task": "2"}),
    "aws_appautoscaling_target.t": Resource("aws_appautoscaling_target.t", {},
                                            lifecycle={"replace_triggered_by":
                                                       ["aws_ecs_service.svc"]}),
}
st = {"aws_ecs_service.svc": {"task": "1"}, "aws_appautoscaling_target.t": {}}
a, e, cbd, order, trig = run(st, cfg, {})
check("F1 引用目标有 update 即触发替换",
      a["aws_appautoscaling_target.t"] == "replace", a)
check("F2 触发原因被记录", trig["aws_appautoscaling_target.t"] == ["aws_ecs_service.svc"], trig)

st2 = {"aws_ecs_service.svc": {"task": "2"}, "aws_appautoscaling_target.t": {}}
a, e, cbd, order, trig = run(st2, cfg, {})
check("F3 目标无变化则不触发", a["aws_appautoscaling_target.t"] == "noop", a)

# 属性级引用
cfg2 = {
    "aws_ecs_service.svc": Resource("aws_ecs_service.svc", {"task": "2"}),
    "t": Resource("t", {}, lifecycle={"replace_triggered_by": ["aws_ecs_service.svc.task"]}),
}
st3 = {"aws_ecs_service.svc": {"task": "1"}, "t": {}}
st4 = {"aws_ecs_service.svc": {"task": "2"}, "t": {}}
a, e, cbd, order, _ = run(st3, cfg2, {})
check("F4 属性级引用按值变否判定", a["t"] == "replace", a)
a, e, cbd, order, _ = run(st4, cfg2, {})
check("F5 属性值未变不触发", a["t"] == "noop", a)

# 官方：只能引用托管资源
cfg3 = {"t": Resource("t", {}, lifecycle={"replace_triggered_by": ["local.x"]})}
a, e, cbd, order, _ = run({"t": {}}, cfg3, {})
check("F6 不能引用 local 值", any("只能引用托管资源" in x for x in e), e)
check("F7 官方触发条件只有 update/replace",
      _update_or_replace("update") and _update_or_replace("replace")
      and not _update_or_replace("create") and not _update_or_replace("noop"))

# ---- 7. precondition 在配置参数求值之前 ----
cfg = {"t": Resource("t", {"a": 1}, lifecycle={"precondition": [
    {"condition": False, "error_message": "nope"}]})}
a, e, cbd, order, _ = run({}, cfg, {})
check("G1 precondition 失败即报错", any("precondition" in x for x in e), e)

# ---- 8. 销毁图与 CBD 翻转 ----
acts = {"a": "replace", "b": "replace"}
edges = destroy_graph({"a": ["b"]}, acts, {"a": False, "b": False})
check("H1 默认先销毁依赖方再销毁被依赖方", edges == [("a", "b")], edges)
edges = destroy_graph({"a": ["b"]}, acts, {"a": True, "b": True})
check("H2 CBD 时销毁边翻转", edges == [("b", "a")], edges)
check("H3 翻转后与另一条边成环",
      has_cycle([("b", "a"), ("a", "b")], ["a", "b"]))
check("H4 无环时不误报", not has_cycle([("a", "b"), ("b", "c")], ["a", "b", "c"]))
edges = destroy_graph({"a": ["b"]}, {"a": "replace", "b": "noop"}, {"a": True, "b": True})
check("H5 对方不销毁则不建边", edges == [], edges)

print("checks=%d fail=%d" % (N, len(FAIL)))
for f in FAIL:
    print("  FAIL", f)

File: 06-DevOps/05-IaC与配置管理/Terraform生命周期元参数与图变换/selfcheck_tflifecycle.go
// tflifecycle 自检：与 Python 版同口径，全部基于实际读过的官方原文。
package main

import (
	"fmt"
	"sort"
	"strings"
)

var n int
var fails []string

func check(label string, cond bool, detail string) {
	n++
	if !cond {
		fails = append(fails, label+"  "+detail)
	}
}

func hasErr(errs []string, sub string) bool {
	for _, e := range errs {
		if strings.Contains(e, sub) {
			return true
		}
	}
	return false
}

func acts(m map[string]string) string {
	ks := make([]string, 0, len(m))
	for k := range m {
		ks = append(ks, k)
	}
	sort.Strings(ks)
	parts := make([]string, 0, len(ks))
	for _, k := range ks {
		parts = append(parts, k+"="+m[k])
	}
	return strings.Join(parts, ",")
}

func main() {
	// ---- 1. 基础动作 ----
	cfg := map[string]Resource{"aws_instance.web": {
		Name: "aws_instance.web", Attrs: map[string]string{"ami": "ami-2"}}}
	r := Plan(map[string]map[string]string{"aws_instance.web": {"ami": "ami-1"}}, cfg, nil)
	check("A1 普通属性变化是 update", r.Actions["aws_instance.web"] == "update" && len(r.Errors) == 0, acts(r.Actions))
	check("A2 默认非 CBD", !r.CBD["aws_instance.web"], "")

	cfg = map[string]Resource{"aws_instance.web": {
		Name: "aws_instance.web", Attrs: map[string]string{"ami": "ami-2"},
		RequiresReplace: map[string]bool{"ami": true}}}
	r = Plan(map[string]map[string]string{"aws_instance.web": {"ami": "ami-1"}}, cfg, nil)
	check("A3 不可原地改的属性→replace", r.Actions["aws_instance.web"] == "replace", acts(r.Actions))
	check("A4 默认先销毁后新建", r.Order["aws_instance.web"] == [2]string{"destroy", "create"}, fmt.Sprint(r.Order))

	// ---- 2. create_before_destroy ----
	cfg = map[string]Resource{"aws_instance.web": {
		Name: "aws_instance.web", Attrs: map[string]string{"ami": "ami-2"},
		RequiresReplace: map[string]bool{"ami": true},
		Lc:              Lifecycle{CBD: boolp(true)}}}
	r = Plan(map[string]map[string]string{"aws_instance.web": {"ami": "ami-1"}}, cfg, nil)
	check("B1 CBD 显式置位", r.CBD["aws_instance.web"], "")
	check("B2 替换次序反过来", r.Order["aws_instance.web"] == [2]string{"create", "destroy"}, fmt.Sprint(r.Order))

	// ---- 3. CBD 传播 ----
	cfg = map[string]Resource{
		"aws_instance.web": {Name: "aws_instance.web", Attrs: map[string]string{"ami": "ami-2"},
			RequiresReplace: map[string]bool{"ami": true}, Lc: Lifecycle{CBD: boolp(true)}},
		"aws_eip.ip": {Name: "aws_eip.ip"},
	}
	deps := map[string][]string{"aws_instance.web": {"aws_eip.ip"}}
	eff, conflicts := PropagateCBD(cfg, deps)
	check("C1 依赖方被隐式置位", eff["aws_eip.ip"], "")
	check("C2 未显式写 false 时无冲突", len(conflicts) == 0, fmt.Sprint(conflicts))

	cfg3 := map[string]Resource{
		"a": {Name: "a", Lc: Lifecycle{CBD: boolp(true)}},
		"b": {Name: "b"}, "c": {Name: "c"},
	}
	eff, _ = PropagateCBD(cfg3, map[string][]string{"a": {"b"}, "b": {"c"}})
	check("C3 传播是传递的", eff["b"] && eff["c"], "")
	eff, _ = PropagateCBD(cfg3, map[string][]string{"b": {"a"}})
	check("C4 不会反向传播给依赖方", eff["a"] && !eff["b"], "")

	cfg4 := map[string]Resource{
		"a": {Name: "a", Lc: Lifecycle{CBD: boolp(true)}},
		"b": {Name: "b", Lc: Lifecycle{CBD: fboolp()}},
	}
	eff, conflicts = PropagateCBD(cfg4, map[string][]string{"a": {"b"}})
	check("C5 显式 false 撞上隐式 true 记为冲突", len(conflicts) == 1 && conflicts[0] == "b", fmt.Sprint(conflicts))
	r = Plan(map[string]map[string]string{}, cfg4, map[string][]string{"a": {"b"}})
	check("C6 冲突会进 errors", hasErr(r.Errors, "成环"), fmt.Sprint(r.Errors))

	// ---- 4. prevent_destroy ----
	cfg = map[string]Resource{"aws_db_instance.db": {
		Name: "aws_db_instance.db", Attrs: map[string]string{"size": "large"},
		RequiresReplace: map[string]bool{"size": true}, Lc: Lifecycle{PreventDestroy: true}}}
	r = Plan(map[string]map[string]string{"aws_db_instance.db": {"size": "small"}}, cfg, nil)
	check("D1 prevent_destroy 拒绝替换", hasErr(r.Errors, "prevent_destroy"), fmt.Sprint(r.Errors))
	r = Plan(map[string]map[string]string{"aws_db_instance.db": {"size": "small"}},
		map[string]Resource{"aws_db_instance.db": {
			Name: "aws_db_instance.db", Attrs: map[string]string{"size": "small"},
			Lc: Lifecycle{PreventDestroy: true}}}, nil)
	check("D2 无变化时放行", r.Actions["aws_db_instance.db"] == "noop" && len(r.Errors) == 0, acts(r.Actions))
	r = Plan(map[string]map[string]string{"aws_db_instance.db": {"size": "small"}}, map[string]Resource{}, nil)
	check("D3 配置移除后照常 destroy", r.Actions["aws_db_instance.db"] == "destroy", acts(r.Actions))
	r = Plan(map[string]map[string]string{}, cfg, nil)
	check("D4 仅创建时不触发 prevent_destroy",
		r.Actions["aws_db_instance.db"] == "create" && len(r.Errors) == 0, acts(r.Actions))

	// ---- 5. ignore_changes ----
	cfg = map[string]Resource{"aws_instance.web": {
		Name: "aws_instance.web", Attrs: map[string]string{"tags": "b"},
		Lc: Lifecycle{IgnoreChanges: []string{"tags"}}}}
	r = Plan(map[string]map[string]string{"aws_instance.web": {"tags": "a"}}, cfg, nil)
	check("E1 被忽略的属性不产生 update", r.Actions["aws_instance.web"] == "noop", acts(r.Actions))
	cfg = map[string]Resource{"aws_instance.web": {
		Name: "aws_instance.web", Attrs: map[string]string{"tags": "b", "ami": "x"},
		Lc: Lifecycle{IgnoreChanges: []string{"tags"}}}}
	r = Plan(map[string]map[string]string{"aws_instance.web": {"tags": "a", "ami": "y"}}, cfg, nil)
	check("E2 只忽略列出的属性", r.Actions["aws_instance.web"] == "update", acts(r.Actions))
	cfg = map[string]Resource{"aws_instance.web": {
		Name: "aws_instance.web", Attrs: map[string]string{"tags": "b"}, Lc: Lifecycle{IgnoreAll: true}}}
	r = Plan(map[string]map[string]string{"aws_instance.web": {"tags": "a"}}, cfg, nil)
	check("E3 all 时永不 update", r.Actions["aws_instance.web"] == "noop", acts(r.Actions))
	r = Plan(map[string]map[string]string{}, cfg, nil)
	check("E4 ignore_changes 不影响 create", r.Actions["aws_instance.web"] == "create", acts(r.Actions))
	cfg = map[string]Resource{"aws_instance.web": {
		Name: "aws_instance.web", Attrs: map[string]string{"ami": "b"},
		RequiresReplace: map[string]bool{"ami": true},
		Lc:              Lifecycle{IgnoreChanges: []string{"ami"}}}}
	r = Plan(map[string]map[string]string{"aws_instance.web": {"ami": "a"}}, cfg, nil)
	check("E5 模型口径：忽略项不进 diff 故不触发 replace",
		r.Actions["aws_instance.web"] == "noop", acts(r.Actions))

	// ---- 6. replace_triggered_by ----
	cfg = map[string]Resource{
		"aws_ecs_service.svc": {Name: "aws_ecs_service.svc", Attrs: map[string]string{"task": "2"}},
		"aws_appautoscaling_target.t": {Name: "aws_appautoscaling_target.t",
			Lc: Lifecycle{ReplaceTriggeredBy: []string{"aws_ecs_service.svc"}}},
	}
	st := map[string]map[string]string{
		"aws_ecs_service.svc": {"task": "1"}, "aws_appautoscaling_target.t": {}}
	r = Plan(st, cfg, nil)
	check("F1 目标有 update 即触发替换", r.Actions["aws_appautoscaling_target.t"] == "replace", acts(r.Actions))
	check("F2 触发原因被记录",
		len(r.Triggers["aws_appautoscaling_target.t"]) == 1 &&
			r.Triggers["aws_appautoscaling_target.t"][0] == "aws_ecs_service.svc", fmt.Sprint(r.Triggers))
	r = Plan(map[string]map[string]string{
		"aws_ecs_service.svc": {"task": "2"}, "aws_appautoscaling_target.t": {}}, cfg, nil)
	check("F3 目标无变化则不触发", r.Actions["aws_appautoscaling_target.t"] == "noop", acts(r.Actions))

	cfg2 := map[string]Resource{
		"aws_ecs_service.svc": {Name: "aws_ecs_service.svc", Attrs: map[string]string{"task": "2"}},
		"t": {Name: "t", Lc: Lifecycle{ReplaceTriggeredBy: []string{"aws_ecs_service.svc.task"}}},
	}
	r = Plan(map[string]map[string]string{"aws_ecs_service.svc": {"task": "1"}, "t": {}}, cfg2, nil)
	check("F4 属性级引用按值变否判定", r.Actions["t"] == "replace", acts(r.Actions))
	r = Plan(map[string]map[string]string{"aws_ecs_service.svc": {"task": "2"}, "t": {}}, cfg2, nil)
	check("F5 属性值未变不触发", r.Actions["t"] == "noop", acts(r.Actions))

	cfg5 := map[string]Resource{"t": {Name: "t",
		Lc: Lifecycle{ReplaceTriggeredBy: []string{"local.x"}}}}
	r = Plan(map[string]map[string]string{"t": {}}, cfg5, nil)
	check("F6 不能引用 local 值", hasErr(r.Errors, "只能引用托管资源"), fmt.Sprint(r.Errors))
	check("F7 官方触发条件只有 update/replace",
		updateOrReplace("update") && updateOrReplace("replace") &&
			!updateOrReplace("create") && !updateOrReplace("noop"), "")

	// ---- 7. precondition ----
	cfg6 := map[string]Resource{"t": {Name: "t", Attrs: map[string]string{"a": "1"},
		Lc: Lifecycle{Preconditions: []Precond{{false, "nope"}}}}}
	r = Plan(map[string]map[string]string{}, cfg6, nil)
	check("G1 precondition 失败即报错", hasErr(r.Errors, "precondition"), fmt.Sprint(r.Errors))

	// ---- 8. 销毁图与 CBD 翻转 ----
	acts1 := map[string]string{"a": "replace", "b": "replace"}
	e1 := DestroyGraph(map[string][]string{"a": {"b"}}, acts1, map[string]bool{"a": false, "b": false})
	check("H1 默认先销毁依赖方再销毁被依赖方", len(e1) == 1 && e1[0] == [2]string{"a", "b"}, fmt.Sprint(e1))
	e1 = DestroyGraph(map[string][]string{"a": {"b"}}, acts1, map[string]bool{"a": true, "b": true})
	check("H2 CBD 时销毁边翻转", len(e1) == 1 && e1[0] == [2]string{"b", "a"}, fmt.Sprint(e1))
	check("H3 翻转后与另一条边成环",
		HasCycle([][2]string{{"b", "a"}, {"a", "b"}}, []string{"a", "b"}), "")
	check("H4 无环时不误报",
		!HasCycle([][2]string{{"a", "b"}, {"b", "c"}}, []string{"a", "b", "c"}), "")
	e1 = DestroyGraph(map[string][]string{"a": {"b"}},
		map[string]string{"a": "replace", "b": "noop"}, map[string]bool{"a": true, "b": true})
	check("H5 对方不销毁则不建边", len(e1) == 0, fmt.Sprint(e1))

	fmt.Printf("checks=%d fail=%d\n", n, len(fails))
	for _, f := range fails {
		fmt.Println("  FAIL", f)
	}
}
